app/models/host/managed.rb

```diff
@@ -373,9 +376,9 @@ def info
               end
 
     info_hash = {}
-    info_hash['classes'] = classes
+    info_hash['classes'] = self.environment ? classes : []
```

Comment:
If you do this a little further up on line 372, then you'll avoid a potentially expensive lookup for Puppet classes which are then skipped.